### PR TITLE
DynamoDB PrimaryKey & Search key information extraction for better query planning

### PR DESCRIPTION
**Pull request by jakobbraun**
The information will be used by `DynamodbQueryRunner` to decide weather a getItem or a scan operation will be used.

### CODE REVIEW

**Comment by jakobbraun** [comment]
According to sonar the code coverage is too low: https://sonarcloud.io/dashboard?branch=keyInfo&id=com.exasol%3Adynamodb-virtual-schema

I did cover the `DynamodbTableMetadataFactory` however with an integration test. Covering it with unit tests makes no sense.

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbKey.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    private final DocumentPathExpression partitionKey;
+    private final DocumentPathExpression sortKey;
+
+    /**
+     * Creates an instance of {@link DynamodbKey}.
+     * 
+     * @param partitionKey partition key of this key
+     * @param sortKey      sort key of this key. {@code null} if no sort key is present.
+     */
+    public DynamodbKey(final DocumentPathExpression partitionKey, final DocumentPathExpression sortKey) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbKey.java:21]: Maybe having two constructors is better than sending null to this one

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadata.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package com.exasol.adapter.dynamodb.dynamodbmetadata;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class describes the key and indexes of a DynamoDB table.
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadata.java:8]:
```suggestion
 * This class describes a key and indexes of a DynamoDB table.
```

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.model.KeySchemaElement;
+import com.amazonaws.services.dynamodbv2.model.TableDescription;
+import com.exasol.adapter.dynamodb.documentpath.DocumentPathExpression;
+
+/**
+ * This class build {@link DynamodbTableMetadata} by fetching the required information using a {@code describeTable}
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:12]:
```suggestion
 * This class builds {@link DynamodbTableMetadata} by fetching the required information using a {@code describeTable}
```

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+public class DynamodbTableMetadataFactory {
+
+    /**
+     * Builds {@link DynamodbTableMetadata} for a DynamoDB table.
+     * 
+     * @param connection DynamoDB connection used for {@code describeTable} call
+     * @param tableName  DynamoDB table name
+     * @return {@link DynamodbTableMetadata} describing the keys and indexes of the table
+     */
+    public DynamodbTableMetadata forDynamodbTable(final AmazonDynamoDB connection, final String tableName) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:24]: maybe `buildMetadataForTable`

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    private DocumentPathExpression extractPartitionKey(final List<KeySchemaElement> keySchema) {
+        for (final KeySchemaElement keySchemaElement : keySchema) {
+            if (keySchemaElement.getKeyType().equals("HASH")) {
+                final String keyName = keySchemaElement.getAttributeName();
+                return new DocumentPathExpression.Builder().addObjectLookup(keyName).build();
+            }
+        }
+        throw new IllegalStateException("Could not find partition key. "
+                + "That's strange because each Dynamodb table must define a partition key.");
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:46]: I don't like this "That's strange" in the exception message :D

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        return null;
+    }
+
+    private List<DynamodbKey> extractLocalSecondaryIndex(final TableDescription tableDescription) {
+        if (tableDescription.getLocalSecondaryIndexes() != null) {
+            return tableDescription.getLocalSecondaryIndexes().stream().map(index -> extractKey(index.getKeySchema()))
+                    .collect(Collectors.toList());
+        } else {
+            return List.of();
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:64]: `Collections.emptyList();` doesn't create a new object

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    private DocumentPathExpression extractSortKey(final List<KeySchemaElement> keySchema) {
+        for (final KeySchemaElement keySchemaElement : keySchema) {
+            if (keySchemaElement.getKeyType().equals("RANGE")) {
+                final String keyName = keySchemaElement.getAttributeName();
+                return new DocumentPathExpression.Builder().addObjectLookup(keyName).build();
+            }
+        }
+        return null;
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:56]: A good candidate to try Optional here. Especially if you add the second constructor as I suggested

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    private List<DynamodbKey> extractLocalSecondaryIndex(final TableDescription tableDescription) {
+        if (tableDescription.getLocalSecondaryIndexes() != null) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:60]: Extract a variable from tableDescription.getLocalSecondaryIndexes()

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            return tableDescription.getLocalSecondaryIndexes().stream().map(index -> extractKey(index.getKeySchema()))
+                    .collect(Collectors.toList());
+        } else {
+            return List.of();
+        }
+    }
+
+    private List<DynamodbKey> extractGlobalSecondaryIndex(final TableDescription tableDescription) {
+        if (tableDescription.getGlobalSecondaryIndexes() != null) {
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:69]: same here

File: src/test/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataTest.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import com.exasol.utils.StringSerializer;
+
+class DynamodbTableMetadataTest {
+    DynamodbKey PRIMARY_KEY = new DynamodbKey(null, null);
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataTest.java:14]: private?

File: src/test/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataTest.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    DynamodbKey GLOBAL_INDEX = new DynamodbKey(null, null);
+    DynamodbKey LOCAL_INDEX = new DynamodbKey(null, null);
+
+    private DynamodbTableMetadata getTableMetadata() {
+        return new DynamodbTableMetadata(this.PRIMARY_KEY, List.of(this.LOCAL_INDEX), List.of(this.GLOBAL_INDEX));
+    }
+
+    @Test
+    void testGetPrimaryKey() {
+        final DynamodbTableMetadata tableMetadata = getTableMetadata();
```
Comment (AnastasiiaSergienko) [inline src/test/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataTest.java:24]: This can go to BeforeEach

Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataTest.java:24]: Or even BeforeAll if you don't change the object

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java (hunk 5 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:46]: better like this?

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java (hunk 7 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:56]: I can't use it inside of `DynamodbKey` as it gets also serialized. This was the class why I asked you ;-) I did however change the constructor to use optional and now only use the null value internal. Do you think this is better?

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java (hunk 6 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:64]:
```
static <E> List<E> of() {
        return ImmutableCollections.emptyList();
    }

static <E> List<E> emptyList() {
        return ImmutableCollections.ListN.EMPTY_LIST;
    }
```

its the same ;-) I also discussed this with sebastian but we agreed `Collections.emptyList();` is better readable

Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbTableMetadataFactory.java:64]: But it's not immutable...

File: src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbKey.java (hunk 1 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/dynamodbmetadata/DynamodbKey.java]: I use this constructor only in `DynamodbTableMetadataFactory`. There it is more practicable to do it with the optional here. I kept it public for the use in other packages tests

**Pull request merged**